### PR TITLE
Fix control vars

### PR DESCRIPTION
**Pull request by gwenchee**

### CODE REVIEW

File: rollo/evaluation.py [hunk 1]
```diff
@@ -705,13 +705,8 @@ def name_ind(self, ind, control_dict, input_evaluators):
         for solver in input_evaluators:
             control_vars[solver] = {}
         for i, var in enumerate(control_dict):
-            if control_dict[var][1] == 1:
-                ind_vars = ind[i]
-            else:
-                ind_vars = []
-                for j in range(control_dict[var][1]):
-                    ind_vars.append(ind[i + j])
-            control_vars[control_dict[var][0]][var] = ind_vars
```
Comment (yardasol) [inline rollo/evaluation.py:714]: I'm also not sure what this change is doing.

File: rollo/toolbox_generator.py [hunk 2]
```diff
@@ -136,9 +136,8 @@ def min_max_list(self, control_dict, input_ctrl_vars):
         min_list = []
         max_list = []
         for var in control_dict:
-            for i in range(control_dict[var][1]):
-                min_list.append(input_ctrl_vars[var]["min"])
-                max_list.append(input_ctrl_vars[var]["max"])
```
Comment (yardasol) [inline rollo/toolbox_generator.py:141]: Ditto

File: tests/unit_tests/input_test_files/input_test_render_jinja_template_python.py [hunk 3]
```diff
@@ -1,2 +1,2 @@
 total_pf = {{packing_fraction}}
-poly_coeff = {{polynomial_triso}}
+variable2 = {{variable2}}
```
Comment (yardasol) [inline tests/unit_tests/input_test_files/input_test_render_jinja_template_python.py:2]: This change specifies that we have a more generic control variable, right?

Comment (yardasol) [reply tests/unit_tests/input_test_files/input_test_render_jinja_template_python.py:2]: This kind of change also exits in `unit_tests/` changes.

File: tests/unit_tests/test_toolbox_generator.py [hunk 4]
```diff
@@ -8,28 +8,20 @@
...
-            "height": 10,
-        },
+        "variable2": {"min": 1, "max": 2},
     },
     "evaluators": {
         "evaluator_1": {
             "order": 0,
-            "input_script": "input_test_eval_fn_generator_template.py",
-            "inputs": ["packing_fraction", "polynomial_triso"],
+            "input_script": "input_test_eval_fn_generator_openmc_template.py",
```
Comment (yardasol) [inline tests/unit_tests/test_toolbox_generator.py:16]: This reverts the a change from #35. Is this on purpose?

**Review by yardasol — CHANGES_REQUESTED** [review]
- [x] Read the PR description
- [x] Read through all the changes, considering the following questions.
  - [ ] Is there anything you can praise about this PR? Start with
        praise.
  - [ ] Are variable names brief but descriptive?
  - [ ] Are new/changed functions no longer than a paragraph?
  - [ ] Do all function parameters have default values where appropriate?
  - [ ] Is the code clear and clean? (see Robert C. Martin's
        [Clean Code](https://i-share-uiu.primo.exlibrisgroup.com/discovery/fulldisplay?docid=alma99944155312205899&context=L&vid=01CARLI_UIU:CARLI_UIU&tab=LibraryCatalog&lang=en))
  - [ ] Is there enough documentation?
  - [ ] Does the programming style meet the requirements of the
        repository ([PEP8](https://www.python.org/dev/peps/pep-0008/) for python,
        [google C++ style guide](https://google.github.io/styleguide/cppguide.html), etc.)
  - [ ] If a new feature has been added, or a bug fixed, has a test been
        added to confirm good behavior?
  - [ ] Does the test actually test the new/changed functionality?
  - [ ] Does the test successfully test edge cases?
  - [ ] Does the test successfully test corner cases?
  - [ ] If the repository has continuous integration: does the PR pass
        the tests?
  - [ ] If there is no continuous integration, check out the branch
        locally, build, and run the tests.
  - [ ] Do the tests pass on your machine?
  - [ ] Is the PR free of random cruft (built files, `.swp` files,
        etc.)?
  - [ ] Do all files in the PR belong in the repository?
  - [ ] If the PR deletes files, is this appropriate?
  - [ ] If the PR adds files or data, are these new files compatible
        with the repository license?
  - [ ] Make a review, leaving kind comments and suggesting changes
        where needed (to resolve the above).
  - [ ] Has the author resolved all of the comments and suggestions
        in your review?
- [ ] When you approve of the PR, merge and close it.
- [ ] Does this PR close an issue? If so, be sure to descriptively
      close this issue when the PR is merged
- [ ] Thank the author for their contribution.

Hi @gwenchee, thanks for your contribution. Glad to see that you're working on fixes to tricky problems like this. It'll make ROLLO more robust going forward.

Could you describe how the changes in this PR address issue #46 in the PR body? That insight is very helpful when reviewing. Right now I'm really not sure what I'm supposed to be looking for. Once you do I'll take another look and give you a more thorough code review.

I noticed that a bunch of your changes from #35 got reverted somewhere. Looks like #37 is the culprit. From the title of #37, I assume you only intended to change the part that touched job parallelization, but it looks like there were other things in that commit. If this is all intentional that's fine too.

I do have some big-picture changes I'd like you to consider:
- [ ] Consider using a more descriptive name than `var` like `control_var` or `c_variable`. `var` on its own is ambiguous.
- [ ] Consider using a more descriptive name than `Ind` like `Individual` or `c_variable`. `var` on its own is ambiguous.
- [ ] There seems to be some interchangeable use of `solver` with `evaluator` and `input` with `control variable`. I'd strongly recommend picking one term and using it consistently everywhere (docs, schema, code, comments).

Also, for my own sanity, could you make your future commit messages more detailed? It would help out a lot when I'm reviewing your PRs if I can browse the commits and see what your motive/intended change is in the commit message, especially when there's lots of refactoring going on.

File: rollo/backend.py [hunk 5]
```diff
@@ -94,16 +94,9 @@ def ind_naming(self):
 
         """
 
-        names = []
-        for ind in self.control_dict:
-            if self.control_dict[ind][1] > 1:
-                for i in range(self.control_dict[ind][1]):
-                    names.append(ind + "_" + str(i))
-            else:
-                names.append(ind)
```
[the remark this answers is not in the document]
Comment (gwenchee) [reply rollo/backend.py:103]: This is some legacy code where I had a nested dictionary in which control_dict[ind][1] had a value that determined whether a variable had multiple values. it was very convoluted so I removed it

File: rollo/evaluation.py (hunk 1 above)
Comment (gwenchee) [reply rollo/evaluation.py:714]: This is the same as above: This is some legacy code where I had a nested dictionary in which control_dict[ind][1] had a value that determined whether a variable had multiple values. it was very convoluted so I removed it. So I removed the if-else statement.

Comment (gwenchee) [reply rollo/evaluation.py:714]: So, if you look at my documentation for control_vars variable: 

```
        control_vars : dict
            multiple layers of dict
            layer 1: solver name
            layer 2: control parameter str
            layer 3: control parameter value
```

This change in lines 708 and 709 adds the relevant control parameters to each nested dictionary for each solver. Thus, ROLLO now accepts the same control variable for different solvers.

File: rollo/toolbox_generator.py (hunk 2 above)
Comment (gwenchee) [reply rollo/toolbox_generator.py:141]: see above

File: tests/unit_tests/input_test_files/input_test_render_jinja_template_python.py (hunk 3 above)
Comment (gwenchee) [reply tests/unit_tests/input_test_files/input_test_render_jinja_template_python.py:2]: exactly

File: tests/unit_tests/test_toolbox_generator.py (hunk 4 above)
Comment (gwenchee) [reply tests/unit_tests/test_toolbox_generator.py:16]: yup

Comment (yardasol) [reply tests/unit_tests/test_toolbox_generator.py:16]: The purpose of #35 was to remove functionality that no longer existed in ROLLO. Did you add the functionality back in?

**Review by yardasol — APPROVED** [review]
Hi @gwenchee, thanks for the clarification. Overall this looks good. I have a suggestion for variable names in `organize_input_output`.

There are various changes in the test suite that I'm having trouble understanding. Could you explain how the test suite changes relate to the rest of the PR?

File: tests/unit_tests/test_toolbox_generator.py (hunk 4 above)
Comment (gwenchee) [reply tests/unit_tests/test_toolbox_generator.py]: You're right. Looking back, I did accidentally change it back. I've committed a change to revert back to `input_test_eval_fn_generator_template.py`. Thanks for catching that.

File: rollo/executor.py [hunk 6]
```diff
@@ -123,22 +123,24 @@ def organize_input_output(self, input_dict):
...
 
         # define control variables dict
         control_vars = OrderedDict()
         for solver in input_evaluators:
             for var in input_evaluators[solver]["inputs"]:
-                control_vars[var] = [solver, 1]
+                if var in control_vars:
+                    control_vars[var].append(solver)
+                else:
+                    control_vars[var] = [solver]
```
[the remark this answers is not in the document]
Comment (gwenchee) [reply rollo/executor.py:143]: I like this. Updating the code to reflect it

**Comment by gwenchee** [comment]
Thanks for the suggested variable name updates @yardasol. As for the test suite changes, 
`input_test_render_jinja_template_python.py`: I updated the variable name to use a more generic naming convention. 
`test_backend.py`: Again, I updated the variable name to use a more generic naming convention. And in lines 37-38, I updated the `control_dict` to follow the change I made in the rollo source code to be {control variable: list of solvers}. The changes in line 78 are to remove the legacy code where I had a nested dictionary in which control_dict[ind][1] had a value that determined whether a variable had multiple values (I don't use this anymore). 
`test_evaluation.py`:  In lines 50-51 and 100-101 and 400-401, I updated the `control_dict` to follow the change I made in the rollo source code to be {control variable: list of solvers}. Then I also updated the variable name to use a more generic naming convention and removed the legacy code. 
`test_executor.py`: In lines 62-63, I updated the `control_dict` to follow the change I made in the rollo source code to be {control variable: list of solvers}. 
`test_toolbox_generator.py`: In lines 11, I removed the legacy code. In lines 17, 24, 88, I used more generic variable name. In line 54, 82-23, 101-102, I updated the `control_dict` to follow the change I made in the rollo source code to be {control variable: list of solvers}.

**Comment by gwenchee** [comment]
@yardasol I updated OpenMC to mamba install, so all tests are passing now. Merge?

File: rollo/executor.py [hunk 7]
```diff
@@ -136,27 +136,27 @@ def organize_input_output(self, input_dict):
         # define control variables dict
         control_vars = OrderedDict()
         for solver in input_evaluators:
-            for var in input_evaluators[solver]["inputs"]:
-                if var in control_vars:
-                    control_vars[var].append(solver)
+            for solver_input in input_evaluators[solver]["inputs"]:
+                if solver_input in control_vars:
+                    control_vars[solver_input].append(solver)
```
Comment (yardasol) [inline rollo/executor.py:141]: : )

File: rollo/executor.py [hunk 8]
```diff
@@ -136,27 +136,27 @@ def organize_input_output(self, input_dict):
...
         # define output variables dict
         output_vars = OrderedDict()
         optimized_variable = input_algorithm["optimized_variable"]
         # find optimized variable
         var_to_solver = {}
         for solver in input_evaluators:
-            for var in input_evaluators[solver]["outputs"]:
-                var_to_solver[var] = solver
+            for solver_output in input_evaluators[solver]["outputs"]:
+                var_to_solver[solver_output] = solver
```
Comment (yardasol) [inline rollo/executor.py:152]: Could you combine this loop with the one on line 138?

**Review by yardasol — APPROVED** [review]
Hi @gwenchee, glad to see the CI is working again. I have one comment I'd like you to consider that I my self didn't catch when I initially reviewed this PR.

File: rollo/executor.py (hunk 8 above)
Comment (gwenchee) [reply rollo/executor.py:152]: I could, but it seems cleaner to me to keep them separated. In the current state, future developers can easily see the separate control variable and output variable logic. If I combine them, it only saves one line of code.

Comment (yardasol) [reply rollo/executor.py:152]: It will save you from needing to loop through `input_evaluators` a second time. Because python is interpreted, each iteration in the loop is executed with extra overhead. Removing that extra loop is less about saving lines of code and more saving on execution time.

Comment (gwenchee) [reply rollo/executor.py:152]: Hey @yardasol , I totally understand where you're coming from. But I don't think the speed up is worth it for the loss in code clarity since the input_evaluators dictionary will never have more than a few elements. 

The logic is as such: 
1) Lines 139-143: First loop through of input_evaluators to initialize the control_vars dictionary 
2) Lines 150-152: Second loop through is initialize a var_to_solver dictionary with the solver output information
3) Lines 153-160: Third loop through to initialize output_vars dictionary to var_to_solver's information, after ensuring the optimized variable is placed first in the output_vars OrderedDict. 

The output variable dict initialization relies on 2) and 3), and so by merging 1) and 2), it muddles clarity. 

We can hop on a call if you want to discuss this more.

Comment (gwenchee) [reply rollo/executor.py:152]: I update test_executor.py to ensure it tests 3) logic.
